### PR TITLE
Run metrics

### PR DESCRIPTION
**Pull request by telegraphroad**
Implemented tests for running metrics. All metrics seem to be running correctly. Need to add 3 adidtional features to metadata (see issues)

### CODE REVIEW

**Comment by telegraphroad** [comment]
@vedran-kasalica metric class and sample column metadata json added. test_model_execution updated with new stuff. thresholding not added yet

File: src/faivor/metrics_api.py [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+        all_metrics = {}
+        
+        # detect appropriate metrics module based on model type
+        metrics_module = (classification_metrics if model_type.lower() == "classification" 
+                         else regression_metrics)
+        
+        for metric in metrics_module.performance:
+            try:
+                result = metric.compute(y_true, y_pred)
+                all_metrics[f"performance.{metric.regular_name}"] = self._convert_to_json_serializable(result)
```
Comment (vedran-kasalica) [inline src/faivor/metrics_api.py:186]: I noticed that `_convert_to_json_serializable` is defined as `@staticmethod`. Calling it using `self` implies an instance method. This is the case for all static methods. For some of them this could be just updated by changing `self.` to `MetricsCalculator.`, however in the case of `_convert_to_json_serializable`, it might be better to move the method to a separate (e.g., `utils` or `data_processing`) module, as the code is not `MetricsCalculator`-specific

File: src/faivor/metrics_api.py [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+            except Exception as e:
+                all_metrics[f"performance.{metric.regular_name}"] = f"Error: {str(e)}"
+        
+        for metric in metrics_module.fairness:
```
Comment (vedran-kasalica) [inline src/faivor/metrics_api.py:190]: These 3 segments might be a good place to refactor the code and provide a method that encapsulates the calculation, abstracting from its type, and then call it for `performance`, `fairness` and `explainability` respectively. That way we the code would abstract from the nesting complexity that the metric calculation entails.

File: src/faivor/metrics_api.py [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+        max_val = np.max(y_pred)
+        
+        # check if values are already in valid prob ([0,1]) range
+        if min_val >= 0 and max_val <= 1:
+            return y_pred, "No transformation needed"
+        
+        #### treatment type #1 1: If values look like logits (large range, both positive and negative)
+        if min_val < -1 or max_val > 2:
+            if verbose:
+                print(f"Values range [{min_val:.4f}, {max_val:.4f}] looks like logits, applying sigmoid")
```
Comment (vedran-kasalica) [inline src/faivor/metrics_api.py:267]: using `print` method is good for dev environment, but for proper logging, I would recommend the `logging` library

File: tests/test_model_execution.py [hunk 4]
```diff
@@ -23,17 +33,457 @@ def test_model_execution(shared_datadir):
...
+    y_true_values = []
+    y_pred_values = []
+    
+    for i, (output, pred) in enumerate(zip(expected_outputs, predictions)):
+        try:
+            # expected output has the required key and is numeric?
+            if isinstance(output, dict) and output_field in output:
+                true_val = output[output_field]
+                if true_val == 'x' or true_val == 'X' or true_val == '':
+                    print(f"Skipping non-numeric expected output at index {i}: {true_val}")
```
Comment (vedran-kasalica) [inline tests/test_model_execution.py:65]: While the code library is expected to log progress, if we make tests granular enough, logging might not be needed. It is often sufficient to log messages on assertions. If the logging is still required, then `logging` might be a better option. The thing is that test logs are usually not interesting, we are more curious about whether the tests pass, and if they do not, the "error message" + the library log is usually sufficient.

File: src/faivor/metrics_api.py [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+        max_val = np.max(y_pred)
+        
+        # check if values are already in valid prob ([0,1]) range
+        if min_val >= 0 and max_val <= 1:
+            return y_pred, "No transformation needed"
+        
+        #### treatment type #1 1: If values look like logits (large range, both positive and negative)
+        if min_val < -1 or max_val > 2:
+            if verbose:
+                logger.log(f"Values range [{min_val:.4f}, {max_val:.4f}] looks like logits, applying sigmoid")
```
Comment (vedran-kasalica) [inline src/faivor/metrics_api.py:255]:
```suggestion
                logger.info(f"Values range [{min_val:.4f}, {max_val:.4f}] looks like logits, applying sigmoid")
```
there was an issue detected by `pytype`, this would resolve it. Hopefully after that, the GitHub action will pass

File: src/faivor/metrics_api.py [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+            transformed = 1 / (1 + np.exp(-y_pred))
+            return transformed, "Applied sigmoid transformation (logits to probabilities)"
+        
+        ### treatment type #2: If values are in a consistent range but shifted/scaled
+        if (max_val - min_val) > 0:
+            if verbose:
+                logger.log(f"Values in range [{min_val:.4f}, {max_val:.4f}], applying min-max scaling")
```
Comment (vedran-kasalica) [inline src/faivor/metrics_api.py:262]:
```suggestion
                logger.info(f"Values in range [{min_val:.4f}, {max_val:.4f}], applying min-max scaling")
```

File: src/faivor/metrics_api.py [hunk 7]
```diff
@@ -0,0 +1,600 @@
...
+            transformed = (y_pred - min_val) / (max_val - min_val)
+            return transformed, "Applied min-max scaling to [0,1] range"
+        
+        ### treatment type #3: Fallback to clipping
+        if verbose:
+            logger.log(f"Values outside [0,1] range, applying clipping")
```
Comment (vedran-kasalica) [inline src/faivor/metrics_api.py:268]:
```suggestion
            logger.info(f"Values outside [0,1] range, applying clipping")
```

**Comment by vedran-kasalica** [comment]
> @vedran-kasalica refactored metrics api with better encapsulation, moved helper functions to utils, test function broken down into smaller tests

Thanks. I made a few suggestions that will hopefully resolve the issue that caused the GitHub action to fail. I will have a look at the rest of the changes a bit later.

File: src/faivor/metrics_api.py [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+        y_true_values = []
+        y_pred_values = []
+        
+        for i, (output, pred) in enumerate(zip(self.expected_outputs, self.predictions)):
+            try:
+                # need to check if expected output has the required key and is numeric
+                if isinstance(output, dict) and self.output_field in output:
+                    true_val = output[self.output_field]
+                    if true_val in ('x', 'X', ''):
+                        continue
```
Comment (vedran-kasalica) [inline src/faivor/metrics_api.py:64]: It might be more intuitive to say if `if true_val not in ('x', 'X', '')` and void using continue

File: src/faivor/metrics_api.py [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+                    
+                    true_float = float(true_val)
+                    pred_float = float(pred)
+                    
+                    valid_indices.append(i)
+                    y_true_values.append(true_float)
+                    y_pred_values.append(pred_float)
+                else:
+                    continue
+            except (ValueError, TypeError):
```
Comment (vedran-kasalica) [inline src/faivor/metrics_api.py:74]:
```suggestion
            except (ValueError, TypeError):
```
The `else` clause can be omited here

File: tests/test_model_execution.py [hunk 10]
```diff
@@ -1,39 +1,249 @@
...
+    
+    calculator = MetricsCalculator(
+        model_metadata=model_metadata,
+        predictions=model_predictions,
+        expected_outputs=expected_outputs,
+        inputs=inputs
+    )
+    return calculator
+
+@pytest.mark.parametrize("model_name", MODEL_NAMES)
```
Comment (vedran-kasalica) [inline tests/test_model_execution.py:101]: this is a nice way to parametrise the tests, we should apply this to the other test files once we have time!

File: tests/test_model_execution.py [hunk 11]
```diff
@@ -1,39 +1,249 @@
...
+import pytest
+import json
+import numpy as np
+
+from faivor.model_metadata import ModelMetadata
+from faivor.parse_data import create_json_payloads
+from faivor.run_docker import execute_model
+from faivor.metrics_api import MetricsCalculator
 
 model:str = "pilot-model_1"
```
Comment (vedran-kasalica) [inline tests/test_model_execution.py:20]:
```suggestion
```
I think that we can delete this one.

**Review by vedran-kasalica — APPROVED** [review]
The changes look good and the tests are well structured, thank you!
